src/EFCore.PG/Query/NpgsqlQueryModelVisitor.cs

```diff
@@ -0,0 +1,133 @@
+// TO PROVIDE MAINTENANCE, SUPPORT, UPDATES, ENHANCEMENTS, OR MODIFICATIONS.
+#endregion
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Linq.Expressions;
+using System.Text;
+using JetBrains.Annotations;
+using Microsoft.EntityFrameworkCore.Query.Expressions.Internal;
+using Microsoft.EntityFrameworkCore.Query.ExpressionVisitors;
+using Microsoft.EntityFrameworkCore.Query.ExpressionVisitors.Internal;
+using Microsoft.EntityFrameworkCore.Utilities;
+using Remotion.Linq;
+using Remotion.Linq.Clauses.Expressions;
+using Remotion.Linq.Clauses.ResultOperators;
+
+namespace Microsoft.EntityFrameworkCore.Query
+{
+    class NpgsqlQueryModelVisitor : RelationalQueryModelVisitor
+    {
+        public NpgsqlQueryModelVisitor(
+            [NotNull] EntityQueryModelVisitorDependencies dependencies,
+            [NotNull] RelationalQueryModelVisitorDependencies relationalDependencies,
+            [NotNull] RelationalQueryCompilationContext queryCompilationContext,
+            [CanBeNull] RelationalQueryModelVisitor parentQueryModelVisitor)
+            : base(dependencies, relationalDependencies, queryCompilationContext, parentQueryModelVisitor)
+        {
+        }
+
+        protected override void OptimizeQueryModel(
+            QueryModel queryModel,
+            bool asyncQuery)
+        {
+            Check.NotNull(queryModel, nameof(queryModel));
+
+            var arrayTranslatingVisitor = new ArrayTranslatingVisitor(this);
+
+            queryModel.TransformExpressions(arrayTranslatingVisitor.Visit);
+
+            base.OptimizeQueryModel(queryModel, asyncQuery);
+        }
+
+        class ArrayTranslatingVisitor : ExpressionVisitorBase
+        {
+            private readonly RelationalQueryModelVisitor _queryModelVisitor;
+
+            internal ArrayTranslatingVisitor(RelationalQueryModelVisitor queryModelVisitor)
+            {
+                _queryModelVisitor = queryModelVisitor;
+            }
+
+            protected override Expression VisitSubQuery(SubQueryExpression expression)
+            {
+                // Prefer the default EF Core translation if one exists
+                var result = base.VisitSubQuery(expression);
+                if (result != null)
+                    return result;
+
+                var subQueryModel = expression.QueryModel;
+                var fromExpression = subQueryModel.MainFromClause.FromExpression;
+
+                var properties = MemberAccessBindingExpressionVisitor.GetPropertyPath(
+                    fromExpression, _queryModelVisitor.QueryCompilationContext, out var qsre);
+
+                if (properties.Count == 0)
+                    return null;
+                var lastProperty = properties[properties.Count - 1];
+                if (lastProperty.ClrType.IsArray)
+                {
+                    // Translate someArray.Length
+                    if (subQueryModel.ResultOperators.First() is CountResultOperator)
```

Comment:
I must still be missing something fundamental about Relinq or something (or maybe we just misunderstood each other a bit). My only purpose in this code is to do server translation (i.e. to translate `.Where(e => e.SomeArray.Length == 2)` into `WHERE array_length("e"."SomeArray", 1) = 2`, I was definitely not trying to optimize the QM itself in any way *before* doing translation to SQL.

I wrote the code you see above (`subQueryModel.ResultOperators.First() is CountResultOperator`) simply because Relinq seems to represent `e.SomeArray.Length` as a SubQueryExpression with a result operator of `Count()`. I originally based that code on the [the way EF Core's SqlTranslatingEV translates `.Contains()` into SQL IN](https://github.com/aspnet/EntityFramework/blob/dev/src/EFCore.Relational/Query/ExpressionVisitors/SqlTranslatingExpressionVisitor.cs#L883).

Anyway... I'm going to merge everything as a translation for now, but I'll be glad to make any changes if you think it can be done better.